### DISCUSSION TITLE
tokio::spawn future cannot be sent between threads safely

### DISCUSSION THREAD

**Discussion by FH0** [tokio-rs/tokio#3886] [Q&A] [answered]
There is no problem with the first code, but it is ugly. The second code will report an error. Is there a way not to write ugly code?

```rust
use std::{sync::RwLock, time::Duration};

#[tokio::main]
async fn main() {
    let timeout = RwLock::new(Duration::from_secs(1));
    tokio::spawn(async move {
        tokio::time::sleep({
            let timeout = timeout.read().unwrap().clone();
            timeout
        })
        .await;
    });
}
```

```rust
use std::{sync::RwLock, time::Duration};

#[tokio::main]
async fn main() {
    let timeout = RwLock::new(Duration::from_secs(1));
    tokio::spawn(async move {
        tokio::time::sleep({
            timeout.read().unwrap().clone()
        })
        .await;
    });
}
```
![image](https://user-images.githubusercontent.com/32861476/123102140-55bb7280-d467-11eb-97d9-e2b6c1abb1c6.png)

**Comment by Darksonn** [comment, marked as answer]
This topic is covered in the [shared state chapter][1] of the Tokio tutorial. The first question you should ask yourself is if you really need a lock at all. For example, could this work?
```Rust
use std::time::Duration;

#[tokio::main]
async fn main() {
    let timeout = Duration::from_secs(1);
    tokio::spawn(async move {
        tokio::time::sleep(timeout).await;
    });
}
```
If you _do_ need a lock for the purpose of having a shared value you can modify from several places, a good way to get cleaner code is to define the following helper struct:
```Rust
use std::time::Duration;
use std::sync::{Arc, RwLock};

#[derive(Clone)]
pub struct SharedDuration {
    inner: Arc<RwLock<Duration>>,
}

impl SharedDuration {
    pub fn new(value: Duration) -> Self {
        Self {
            inner: Arc::new(RwLock::new(value)),
        }
    }
    pub fn set(&self, value: Duration) {
        let mut lock = self.inner.write().unwrap();
        *lock = value;
    }
    pub fn get(&self) -> Duration {
        let lock = self.inner.read().unwrap();
        *lock
    }
}

#[tokio::main]
async fn main() {
    let timeout = SharedDuration::new(Duration::from_secs(1));
    tokio::spawn(async move {
        tokio::time::sleep(timeout.get()).await;
    });
}
```
This value can be cloned to get a new handle to the same shared duration. If you call `set` on it, any future calls to `get` will return that new value, and this includes any other cloned handles, even if they were cloned off before the call to `set`.

This technique is described under the [Restructure your code to not hold the lock across an `.await`][2] heading in the chapter linked earlier.

[1]: https://tokio.rs/tokio/tutorial/shared-state
[2]: https://tokio.rs/tokio/tutorial/shared-state#restructure-your-code-to-not-hold-the-lock-across-an-await

**Comment by FH0** [comment]
Thank you for answering such a simple question in detail. I got the answer.
